controllers/topology/current_state.go

```diff
@@ -18,12 +18,142 @@ package topology
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.ControlPlaneRef.Kind, cluster.Spec.ControlPlaneRef.Name)
+	}
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the ClusterClass of the given Cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries for all MachineDeployments and filters them for their linked Cluster and
+// whether they are managed by a ClusterClass using labels. A Cluster may have zero or more MachineDeployments. Zero is
+// expected on first reconcile. If MachineDeployments are found for the Cluster their Infrastructure and Bootstrap references
+// are inspected. Where these are not found the function will throw an error.
+func (r *ClusterReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *clusterv1.Cluster) (map[string]machineDeploymentTopologyState, error) {
+	mDeploymentState := make(map[string]machineDeploymentTopologyState)
+
+	md := &clusterv1.MachineDeploymentList{}
+	err := r.Client.List(ctx, md, client.MatchingLabels{
+		clusterv1.ClusterLabelName:         cluster.Name,
+		clusterv1.ClusterTopologyLabelName: "",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read MachineDeployments for managed topology")
+	}
+	for _, m := range md.Items {
+		machineDeployment := m
+		mdTopologyName, ok := m.ObjectMeta.Labels[clusterv1.ClusterTopologyMachineDeploymentLabelName]
+		if !ok {
+			return nil, fmt.Errorf("failed to find label %s in %s", clusterv1.ClusterTopologyMachineDeploymentLabelName, machineDeployment.Name)
+		}
+		if mdTopologyName == "" {
+			return nil, fmt.Errorf("no value for MachineDeploymentTopoplogyName for %s", machineDeployment.Name)
+		}
+		if _, ok := mDeploymentState[mdTopologyName]; ok {
```

Comment:
@sbueringer this is my understanding of the check if the topology name is in the map.

Reply:
looks great!